docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,951 @@
+DeleteObjects request, it is still useful to send these through the last part of the
+deletion pipeline to coalesce with other executing deletions: to enable this, the
+deletion queue should expose two input channels: one for deletions that must be
+processed in a generation-aware way, and a fast path for timeline deletions, where
+that fast path may skip validation and the persistent queue.
+
+### Cleaning up orphan objects (scrubbing)
+
+An orphan object is any object which is no longer referenced by a running node or by metadata.
+
+Examples of how orphan objects arise:
+
+- A node PUTs a layer object, then crashes before it writes the
+  index_part.json that references that layer.
+- A stale node carries on running for some time, and writes out an unbounded number of
+  objects while it believes itself to be the rightful writer for a tenant.
+- A pageserver crashes between un-linking an object from the index, and persisting
+  the object to its deletion queue.
+
+Orphan objects are functionally harmless, but have a small cost due to S3 capacity consumed. We
+may clean them up at some time in the future, but doing a ListObjectsv2 operation and cross
+referencing with the latest metadata to identify objects which are not referenced.
+
+Scrubbing will be done only by an attached pageserver (not some third party process), and deletions requested during scrub will go through the same
+validation as all other deletions: the attachment generation must be
+fresh. This avoids the possibility of a stale pageserver incorrectly
+thinking than an object written by a newer generation is stale, and deleting
+it.
+
+## Operational impact
+
+### Availability
+
+Coordination of generation numbers via the control plane introduce a dependency for certain
+operations:
+
+1. Starting new pageservers (or activating pageservers after a restart)
+2. Executing enqueued deletions
+3. Advertising updated `remote_consistent_lsn` to enable WAL trimming
+
+Item 1. would mean that some in-place restarts that previously would have resumed service even if the control plane were
+unavailable, will now not resume service to users until the control plane is available. We could
+avoid this by having a timeout on communication with the control plane, and after some timeout,
+resume service with the node's previous generation (assuming this was persisted to disk). However,
+this is unlikely to be needed as the control plane is already an essential & highly available component. Also, having a node re-use an old generation number would complicate
+reasoning about the system, as it would break the invariant that one (node_id, generation)
+tuple uniquely identifies one process lifetime -- it is not recommended to implement this.
+
+Item 2. is a non-issue operationally: it's harmless to delay deletions, the only impact of objects pending deletion is
+the S3 capacity cost.
+
+Item 3. is an issue if safekeepers are low on disk space and the control plane is unavailable for a long time.
```

Comment:
This can be mitigated by changing safekeepers to allow offloading of WAL to S3 _prior_ to pruning.

I haven't looked at Safekeeper code at all, maybe they support it already.

It'd basically be the equivalent of on-demand downloads but in Safekeepers.

Also, to improve WAL durability beyond what SKs in 3 different AZs can offer I think it makes sense for Safekeepers to upload WAL to S3 ASAP anyways.

Let's ask the team about this.

Reply:
https://neondb.slack.com/archives/C04KGFVUWUQ/p1692967665177109

Reply:
(soften the description of safekeeper impact, or just remove it as this RFC isn't meant to define the safekeeper behavior)

Reply:
Safekeepers already upload segment ASAP once it is filled. As discussed in thread above, we can implement deleting local segments if they are already uploaded and pageserver lags too much. So proposed is fine from here.